WeCantSpell.Hunspell/WordList.Query.cs

```diff
@@ -937,203 +917,60 @@ rv is not null
-                                    if (CompoundWordPairCheck(wordLenPrefix))
-                                    {
-                                        return null;
-                                    }
-
-                                    if (Affix.CheckCompoundRep || Affix.ForbiddenWord.HasValue)
-                                    {
-                                        if (Affix.CheckCompoundRep && CompoundReplacementCheck(wordLenPrefix))
+                                        // forbid compound word, if it is a non-compound word with typical
+                                        // fault, or a dictionary word pair
+                                        switch (CompoundCheckDecideForbidFinal(word, len, i, st, rv))
                                         {
-                                            st.Destroy();
-                                            return null;
-                                        }
-
-                                        // check first part
-                                        if (i < word.Length && word.Slice(i).StartsWith(rv.Word.AsSpan()))
-                                        {
-                                            var r = st.Exchange(i + rv.Word.Length, '\0');
-
-                                            var stString = st.TerminatedSpan;
-                                            if ((Affix.CheckCompoundRep && CompoundReplacementCheck(stString)) || CompoundWordPairCheck(stString))
-                                            {
-                                                st[i + rv.Word.Length] = r;
-
-                                                continue;
-                                            }
-
-                                            if (Affix.ForbiddenWord.HasValue)
-                                            {
-                                                var rv2 = LookupFirst(word);
-                                                if (rv2 is null && len <= word.Length)
-                                                {
-                                                    rv2 = AffixCheck(word.Slice(0, len), default, CompoundOptions.Not);
-                                                }
-
-                                                if (
-                                                    rv2 is not null
-                                                    && rv2.ContainsFlag(Affix.ForbiddenWord)
-                                                    && equalsOrdinalLimited(rv2.Word.AsSpan(), st.TerminatedSpan, i + rv.Word.Length)
-                                                )
-                                                {
-                                                    st.Destroy();
-                                                    return null;
-                                                }
-                                            }
+                                            case CompoundCheckOutcomes.Fail:
```

Comment:
Make sure these cases line up with the enum ordering so the compiler has more opportunity to do cool jump tricks.